src/btree/bt_sync.c

```diff
@@ -56,13 +56,19 @@ __sync_file(WT_SESSION_IMPL *session, int syncop)
 
 		flags |= WT_READ_NO_WAIT | WT_READ_SKIP_INTL;
 		for (walk = NULL;;) {
-			WT_ERR(__wt_tree_walk(session, &walk, flags));
+			WT_ERR(__wt_tree_walk(session, &walk, NULL, flags));
 			if (walk == NULL)
 				break;
 
-			/* Write dirty pages if nobody beat us to it. */
+			/*
+			 * Write dirty pages if nobody beat us to it.  Don't
+			 * try to write the hottest pages: checkpoint will have
+			 * to visit them anyway.
+			 */
 			page = walk->page;
-			if (__wt_page_is_modified(page)) {
+			if (__wt_page_is_modified(page) &&
+			    __wt_txn_visible_all(
+			    session, page->modify->update_txn)) {
 				if (txn->isolation == TXN_ISO_READ_COMMITTED)
 					__wt_txn_refresh(session, 1);
 				leaf_bytes += page->memory_footprint;
```

Comment:
Maybe I can update the comment to make this clearer: if the page is modified and the updates are old enough (i.e., all visible), then we try to write it as part of write-leaves.  If the updates are too recent (i.e., some are not visible), we skip it during write-leaves because we're going to have to write it during the checkpoint phase.

Does that make sense?